sycl/include/sycl/interop_handle.hpp

```diff
@@ -182,6 +187,66 @@ class interop_handle {
 #endif
   }
 
+  // Adds events from the native API so that the SYCL runtime can integrate the
+  // async calls in a native API, with other async operations in the SYCL DAG.
+  // Using this function removes the requirement that a host task callable must
+  // synchronize with any asynchronous operations from within the callable.
+  template <backend Backend = backend::opencl>
+  void add_native_events(backend_return_t<Backend, event> NativeEvents) {
+#ifndef __SYCL_DEVICE_ONLY__
+    // TODO: replace the exception thrown below with the SYCL 2020 exception
+    // with the error code 'errc::backend_mismatch' when those new exceptions
+    // are ready to be used.
+    if (Backend != get_backend())
+      throw invalid_object_error("Incorrect backend argument was passed",
+                                 PI_ERROR_INVALID_MEM_OBJECT);
+    // All native events can be cast to void*, we use this as a generic entry
+    // point to source library
+    std::vector<pi_native_handle> NativeEventHolders(NativeEvents.size());
+    for (auto i = 0; i < NativeEvents.size(); ++i)
+      NativeEventHolders[i] =
+          reinterpret_cast<pi_native_handle>(NativeEvents[i]);
+    return addNativeEvents(NativeEventHolders);
+#else
+    // we believe this won't be ever called on device side
+    return;
+#endif
+  }
+
+  // Gets all the native events that the host task depends on, and that are
+  // still active
+  template <backend Backend = backend::opencl>
+  backend_return_t<Backend, event> get_native_events() {
```

Comment:
Yes maybe this is a good idea. The problem is that `backend_return_t` is already a `std::vector` for many backends. Allocating a vector of vectors would be unnecessarily bad in terms of perf overhead. So maybe a `std::vector<backend_input_t<...>>` is better?

Reply:
We want a vector here regardless of whether `backend_return_t` already is a vector or not. Using `backend_input_t` in a return type feels wrong, although I think it would work just fine. A third option would be to use some template magic to make a vector out of `backend_return_t` if it is not already a vector.